File: backend/src/routers/user_collections.py
# 顶层模块（imports 与 router 定义）
from fastapi import APIRouter, Depends, HTTPException, Path, Query, Body, Request
from typing import List, Optional
from pydantic import BaseModel, Field, constr
from src.models import (
    CustomListUpdate,
    CustomListCreate,
    CustomListInDB,
    CustomListPageResult,
    CustomListRead,
    CustomListType,
    MovieRead,
)
from src.routers.schemas.movie import MovieReadResponseSchema
from src.routers.schemas.user_custom_list import (
    CustomListCreateRequestSchema,
    CustomListUpdateRequestSchema,
    CustomListReadResponseSchema,
    CustomListPageResultResponseSchema,
    AddMoviesSchema, RemovieMoviesSchema,
)
from src.services import CollectionService
from src.core.dependencies import get_current_user  # 返回当前登录用户对象
from src.core.handler import router_handler
from config.logging import get_router_logger, set_trace_id, get_trace_id

router = APIRouter(prefix="/api/v1/user_collections", tags=["User_Collections"])
service = CollectionService()
logger = get_router_logger("user_collections")


# -----------------------------
# 路由层实现
# -----------------------------
@router.get("/", response_model=CustomListPageResultResponseSchema)
@router_handler(action="list_collections")
async def get_user_collections(
    request: Request,
    current_user=Depends(get_current_user),
    type: Optional[CustomListType] = Query(
        None, description="可选合集类型筛选: favorite, watchlist, custom"
    ),
    query: Optional[str] = Query(None, description="关键词模糊搜索"),
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=200, description="每页数量"),
):
    return await service.list_collections(
        current_user.id,
        type_filter=type,
        query=query,
        page=page,
        page_size=size,
        current_user=current_user,
    )


@router.get("/{collection_id}", response_model=CustomListReadResponseSchema)
@router_handler(action="get_collection_detail")
async def get_collection_detail(
    request: Request,
    collection_id: str = Path(..., description="合集ID"),
    current_user=Depends(get_current_user),
):
    collection = await service.get_collection(collection_id, current_user=current_user)
    if not collection:
        raise HTTPException(status_code=404, detail="合集不存在或无权限访问")
    return collection


@router.post("/", response_model=CustomListReadResponseSchema)
@router_handler(action="create_collection")
async def create_collection(
    request: Request,
    data: CustomListCreateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    if data.type not in CustomListType.__members__.values():
        raise HTTPException(status_code=400, detail="非法合集类型")
    return await service.create_collection(data, current_user=current_user)

@router.post("/initUserDefaultCollections", response_model=List[CustomListReadResponseSchema])
@router_handler(action="create_collection")
async def create_collection(
    request: Request,
    current_user=Depends(get_current_user),
):
    return await service.init_user_default_collections(current_user=current_user)

@router.patch("/{collection_id}", response_model=CustomListReadResponseSchema)
@router_handler(action="update_collection")
async def update_collection(
    request: Request,
    collection_id: str = Path(..., description="合集ID"),
    data: CustomListUpdateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    if data.type is not None and data.type not in CustomListType.__members__.values():
        raise HTTPException(status_code=400, detail="非法合集类型")
    return await service.update_collection(
        collection_id, data, current_user=current_user
    )


@router.delete("/{collection_id}", response_model=CustomListReadResponseSchema)
@router_handler(action="delete_collection")
async def delete_collection(
    request: Request,
    collection_id: str = Path(..., description="合集ID"),
    current_user=Depends(get_current_user),
):
    deleted = await service.delete_collection(collection_id, current_user=current_user)
    if not deleted:
        raise HTTPException(status_code=404, detail="合集不存在或已删除")
    return deleted


@router.post("/{collection_id}/add_movies", response_model=dict)
@router_handler(action="add_movies_to_collection")
async def add_movies_to_collection(
    request: Request,
    collection_id: str = Path(..., description="合集ID"),
    data: AddMoviesSchema = Body(...),
    current_user=Depends(get_current_user),
):
    if not data.movie_ids:
        raise HTTPException(status_code=400, detail="影片ID列表不能为空")
    return await service.add_movies(
        collection_id, data.movie_ids, current_user=current_user
    )


@router.post("/{collection_id}/remove_movies", response_model=dict)
@router_handler(action="remove_movies_from_collection")
async def remove_movies_from_collection(
    request: Request,
    collection_id: str = Path(..., description="合集ID"),
    data: RemovieMoviesSchema = Body(...),
    current_user=Depends(get_current_user),
):
    if not data.movie_ids:
        raise HTTPException(status_code=400, detail="影片ID列表不能为空")
    return await service.remove_movies(
        collection_id, data.movie_ids, current_user=current_user
    )

@router.get("/{collection_id}/movies", response_model=List[MovieReadResponseSchema])
@router_handler(action="get_collection_movies")
async def get_collection_movies(
    request: Request,
    collection_id: str = Path(..., description="合集ID"),
    current_user=Depends(get_current_user),
):
    movies = await service.get_collection_movies(collection_id, current_user=current_user)
    return movies

File: frontend/src/api/user_collections.ts
import { http } from "./http";
import {
  CustomListType,
  type CustomListRead,
  type CustomListCreateRequestSchema,
  type CustomListUpdateRequestSchema,
  type CustomListPageResultResponseSchema,
} from "@/types/user_collection";
import type { MovieRead } from "@/types/movie";

/**
 * 前后端 type 值映射：CUSTOMLIST <-> custom
 */
// 前后端 type 值映射：CUSTOMLIST <-> custom
function mapFrontendTypeToBackend(t?: CustomListType | null): string | undefined {
  if (t === CustomListType.CUSTOMLIST) return "customlist";
  if (t === CustomListType.FAVORITE) return "favorite";
  if (t === CustomListType.WATCHLIST) return "watchlist";
  return t as unknown as string | undefined;
}

function mapBackendTypeToFrontend(s: string): CustomListType {
  switch (s) {
    case "customlist":
    case "custom": // 兼容旧值
      return CustomListType.CUSTOMLIST;
    case "favorite":
      return CustomListType.FAVORITE;
    case "watchlist":
      return CustomListType.WATCHLIST;
    default:
      return s as unknown as CustomListType;
  }
}

function mapReadFromBackend(r: any): CustomListRead {
  return {
    ...r,
    type: mapBackendTypeToFrontend(r?.type),
  };
}

/**
 * 列举当前用户的片单：GET /api/v1/user_collections/?type=...
 * 返回分页结果 CustomListPageResultResponseSchema
 */
export async function listUserCollections(
  token: string,
  type?: CustomListType,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<CustomListPageResultResponseSchema> {
  const qp = new URLSearchParams();
  if (type !== undefined && type !== null) {
    const t = mapFrontendTypeToBackend(type);
    if (t) qp.set("type", t);
  }
  const res = await http.get<CustomListPageResultResponseSchema>("/api/v1/user_collections/", {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
  return {
    ...res,
    items: res.items.map(mapReadFromBackend),
  };
}

/**
 * 获取片单详情：GET /api/v1/user_collections/{collection_id}
 */
export async function getUserCollection(
  token: string,
  collectionId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<CustomListRead> {
  const res = await http.get<CustomListRead>(`/api/v1/user_collections/${collectionId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return mapReadFromBackend(res);
}

/**
 * 创建片单：POST /api/v1/user_collections/
 */
export async function createUserCollection(
  token: string,
  payload: CustomListCreateRequestSchema,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<CustomListRead> {
  const payloadForBackend = {
    ...payload,
    type: payload?.type ? mapFrontendTypeToBackend(payload.type) : payload?.type,
  };
  const res = await http.post<CustomListRead>("/api/v1/user_collections/", payloadForBackend, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return mapReadFromBackend(res);
}

/**
 * 更新片单：PATCH /api/v1/user_collections/{collection_id}
 * 注意：后端 CollectionUpdateSchema 仅强约束 name，其他字段取自 CustomListUpdate
 */
export async function updateUserCollection(
  token: string,
  collectionId: string,
  patch: CustomListUpdateRequestSchema,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<CustomListRead> {
  const patchForBackend = {
    ...patch,
  };
  const res = await http.patch<CustomListRead>(`/api/v1/user_collections/${collectionId}`, patchForBackend, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return mapReadFromBackend(res);
}

/**
 * 删除片单：DELETE /api/v1/user_collections/{collection_id}
 */
export async function deleteUserCollection(
  token: string,
  collectionId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<CustomListRead> {
  const res = await http.delete<CustomListRead>(`/api/v1/user_collections/${collectionId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return mapReadFromBackend(res);
}

/**
 * 向片单添加影片：POST /api/v1/user_collections/{collection_id}/add_movies
 * Body: { movie_ids: string[] }
 */
export async function addMoviesToCollection(
  token: string,
  collectionId: string,
  movieIds: string[],
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<Record<string, unknown>> {
  const body = { movie_ids: movieIds };
  return http.post<Record<string, unknown>>(
    `/api/v1/user_collections/${collectionId}/add_movies`,
    body,
    {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
    }
  );
}

/**
 * 从片单移除影片：POST /api/v1/user_collections/{collection_id}/remove_movies
 * Body: { movie_ids: string[] }
 */
export async function removeMoviesFromCollection(
  token: string,
  collectionId: string,
  movieIds: string[],
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<Record<string, unknown>> {
  const body = { movie_ids: movieIds };
  return http.post<Record<string, unknown>>(
    `/api/v1/user_collections/${collectionId}/remove_movies`,
    body,
    {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
    }
  );
}

/**
 * 获取片单影片：GET /api/v1/user_collections/{collection_id}/movies
 * 返回 List[MovieRead]
 */
export async function getCollectionMovies(
  token: string,
  collectionId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<MovieRead[]> {
  return http.get<MovieRead[]>(`/api/v1/user_collections/${collectionId}/movies`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}